source/code/plugin/filter_cadvisor2mdm.rb

```diff
@@ -0,0 +1,214 @@
+
+# frozen_string_literal: true
+
+module Fluent
+    require 'logger'
+    require 'json'
+
+	class CAdvisor2MdmFilter < Filter
+		Fluent::Plugin.register_filter('filter_cadvisor2mdm', self)
+		
+		config_param :enable_log, :integer, :default => 0
+        config_param :log_path, :string, :default => '/var/opt/microsoft/omsagent/log/filter_cadvisor2mdm.log'
+        config_param :custom_metrics_azure_regions, :string
+        config_param :metrics_to_collect, :string, :default => 'cpuUsageNanoCores,memoryWorkingSetBytes,memoryRssBytes'
+        
+        @@cpu_usage_milli_cores = 'cpuUsageMilliCores'
+        @@cpu_usage_nano_cores = 'cpuusagenanocores'
+        @@object_name_k8s_node = 'K8SNode'
+        @process_incoming_stream = true
+        @metrics_to_collect_hash = {}
+		def initialize
+			super
+		end
+
+		def configure(conf)
+			super
+			@log = nil
+			
+			if @enable_log
+				@log = Logger.new(@log_path, 'weekly')
+				@log.debug {'Starting filter_cadvisor2mdm plugin'}
+			end
+		end
+
+        def start
+            super
+            @@custom_metrics_template = '
+            { 
+                "time": "%{timestamp}", 
+                "data": { 
+                    "baseData": { 
+                        "metric": "%{metricName}", 
+                        "namespace": "Insights.Containers/node", 
+                        "dimNames": [ 
+                        "host"
+                        ], 
+                        "series": [ 
+                        { 
+                            "dimValues": [ 
+                            "%{hostvalue}"
+                            ], 
+                            "min": %{metricminvalue},
+                            "max": %{metricmaxvalue}, 
+                            "sum": %{metricsumvalue}, 
+                            "count": 1 
+                        } 
+                        ] 
+                    } 
+                } 
+            }'
+
+            @process_incoming_stream = check_custom_metrics_availability
+            @metrics_to_collect_hash = build_metrics_hash
+            @log.debug "After check_custom_metrics_availability process_incoming_stream #{@process_incoming_stream}"
+
+            # initialize cpu and memory limit 
+            if @process_incoming_stream
+                @cpu_limit = 0.0
+                @memory_limit = 0.0 
+                
+                begin 
+                    nodeInventory = JSON.parse(KubernetesApiClient.getKubeResourceInfo('nodes').body)
+                rescue Exception => e
+                    @log.info "Error when getting nodeInventory from kube API. Exception: #{e.class} Message: #{e.message} "
+                    ApplicationInsightsUtility.sendExceptionTelemetry(e.backtrace)
+                end
+                if !nodeInventory.nil? 
+                    cpu_limit_json = KubernetesApiClient.parseNodeLimits(nodeInventory, "capacity", "cpu", "cpuCapacityNanoCores")
```

Comment:
Can you also do a quick check on the Ux repo, that the Ux is using the 'cpuCapacityNanoCores' and not 'cpuAllocatableNanoCores' to calculate % ?

Reply:
UX used cpuCapacityNanoCores